### ISSUE TITLE
Loss function with "zero-padding"

### ISSUE THREAD

**Issue by EvanBeal** [ivadomed/multimodal-registration#1]
## Context

It has been observed that the registration models trained starting from label maps generated from noise distribution have difficulties to register the moving volumes to the fixed ones in the areas at the borders of zero-padded regions. This is probably due to the fact that being trained with label maps generated from noise distribution, the model is not learning any notion of foreground/background leading to some extreme deformations when zero-padding is present (large background areas).

This behaviour can be observed on this example: 
<img width="1152" alt="Capture d’écran 2021-11-23 à 09 50 46" src="https://user-images.githubusercontent.com/32447627/143046688-c64994f7-de44-4485-8e9c-e10ef3e8a802.png">
The results were obtained using this [model](https://surfer.nmr.mgh.harvard.edu/ftp/data/voxelmorph/synthmorph/shapes-dice-vel-3-res-8-16-32-256f.h5), with this [T1w volume](https://github.com/spinalcordtoolbox/sct_tutorial_data/raw/master/single_subject/data/t1/t1.nii.gz) and this [T2w volume](https://github.com/spinalcordtoolbox/sct_tutorial_data/blob/master/single_subject/data/t2/t2.nii.gz) (of the same patient).
 
Therefore, to try to adapt the registration model to the zero-padded areas, the feature `zero_borders_maps` present in the config file can be set to `true`, in order to train a registration model (using `train_synthmorph.py`) using label maps on which zero-padding is applied.

## Issue description

The current problem is that the loss function is unchanged in this scenario, compared to when no zero-padding is applied. Therefore, the loss is not taking into account this specific case leading to a high and noisy training loss that is difficult to interpret. This implies that no conclusion on how well the model has learned to register the volumes can be done during the training phase.

Example of training loss obtained when setting `zero_borders_maps` to `true`:

<img width="400" alt="Capture d’écran 2021-11-23 à 10 29 14" src="https://user-images.githubusercontent.com/32447627/143053597-c9290266-d3ae-4012-9274-36d1d0819ee1.png">

Compared to when `zero_borders_maps` is set to `false`:

<img width="400" alt="Capture d’écran 2021-11-23 à 10 30 57" src="https://user-images.githubusercontent.com/32447627/143053862-76abd13d-e3e4-446d-953b-6df7a972b104.png">

The results were obtained using the following config file (with the only difference being for the `zero_borders_maps` param):
<details>
  <summary>Config</summary>
  
```
{
  "model_dir": "model",
  "log_dir": "logs",
  "bool_sub_dir": false,
  "sub_dir": "train_ex",
  "gen_label_only": false,
  "gen_label": true,
  "save_label": true,
  "label_dir": "labels",
  "zero_borders_maps": XXX,
  "zero_borders_maps_val": false,
  "in_shape": [128, 128, 128],
  "num_labels": 26,
  "num_maps": 100,
  "im_scales": [16, 32, 64],
  "def_scales": [8, 16, 32],
  "im_max_std": 1,
  "def_max_std": 3,
  "add_str": "26lab_",
  "same_subj": true,
  "blur_std": 1,
  "gamma": 0.25,
  "vel_std": 3,
  "vel_res": 16,
  "bias_std": 0.3,
  "bias_res": 40,
  "gpu": "0",
  "epochs": 1600,
  "batch_size": 4,
  "train_frac": 1,
  "batch_size_val": 4,
  "save_freq": 200,
  "bool_init_weights": false,
  "init_weights": "model.h5",
  "reg_param": 1,
  "lr": 1e-4,
  "init_epoch": 0,
  "verbose": 1,
  "int_steps": 5,
  "int_res": 2,
  "svf_res": 2,
  "enc": [64, 64, 64, 64],
  "dec": [64, 64, 64, 64, 64, 64]
}
```
</details>

## Potential solutions

Ideally, the loss should ignore the registration accuracy in the areas that have been zero-padded and only focus on and penalise the mismatch of the different labels in the areas where there is no zero-padding in the fixed and moving volumes.
The current loss being a Dice loss, it should be adapted to ignore the areas where the voxel values are 0 for one of the two volumes and only compute the volume overlap of the labels (except the 0 label) in the other areas. Doing so would probably guide the model to only register the "foreground" labels, ignoring the zero-padded regions.

**Comment by jcohenadad** [comment]
> The current loss being a Dice loss, it should be adapted to ignore the areas where the voxel values are 0 for one of the two volumes and only compute the volume overlap of the labels (except the 0 label) in the other areas.

@charleygros @andreanne-lemay @mariehbourget I vaguely remember you've already bumped into a similar issue

**Comment by andreanne-lemay** [comment]
From my recollection, the problem we faced was computing Dice on 2 empty volumes (i.e., return 1 or nan in this case). Here, from my understanding, we want to compute the Dice only on pixels that have data on both images. 
A solution to this could be to:
1) Flatten and filter the pixels that have non zero values on both channels by removing them.
2) Compute loss on these new potentially smaller tensors.

**Comment by EvanBeal** [comment]
> A solution to this could be to:
> 
> 1. Flatten and filter the pixels that have non zero values on both channels by removing them.
> 2. Compute loss on these new potentially smaller tensors.

Thanks @andreanne-lemay ! I agree, this seems like a good way to start.
